lib/Dialect/FIRRTL/Import/FIRParser.cpp

```diff
@@ -814,6 +817,33 @@ ParseResult FIRParser::parseType(FIRRTLType &result, const Twine &message) {
     result = ClockType::get(getContext());
     break;
 
+  case FIRToken::kw_Inst: {
+    if (FIRVersion::compare(version, {3, 1, 0}) < 0)
+      return emitError() << "unexpected token: Inst types are a FIRRTL 3.1.0+ "
+                            "feature, but the specified FIRRTL version was "
+                         << version;
+
+    consumeToken(FIRToken::kw_Inst);
+    if (parseToken(FIRToken::less, "expected < in Inst type"))
+      return failure();
+
+    auto loc = getToken().getLoc();
+    StringRef id;
+    if (parseId(id, "expected class name in Inst type"))
+      return failure();
+
+    // Look up the class that is being referenced.
+    auto classOp = getConstants().classMap[id];
```

Comment:
Can we use `classMap.find(id)` since `classMap[id]` mutates `classMap` so it's not thread safe?

Reply:
As we keep adding top-level things... not for this PR since this all seems pre-existing, but:

1)StringMap seems unnecessary since the strings are already stored elsewhere, but shrug.
2)Why a per-op lookup, since these all share the same namespace? This would help catch collisions but also allow more helpful errors should these conflict.  (should this just be `circuit.lookup(id)` or is there some ordering issue such that that's not reasonable?)
3)None of these maps are constants, and don't belong in `SharedParserConstants`, which seems to be for storing cached "constants".  These are supposed to all be initialized once (and immutable thereafter, effectively should be `const`), not populated by one parser layer for use by another....

Reply:
Changed the classMap to a DenseMap<StringRef, ClassOp>. I agree that this is an abuse of the constants. The type parsing is implemented in the base FIRParser, which doesn't hold a pointer to the class op. We could keep a symbol table up to date as we parse the toplevel definitions of the circuit, and pass that down into the subparsers, maybe I can do that as a follow up PR?